staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -160,3 +172,70 @@ func decorateResponseWriter(responseWriter http.ResponseWriter, out io.Writer, i
 	}
 	return delegate
 }
+
+var _ http.ResponseWriter = &auditResponseWriter{}
+var cloner = conversion.NewCloner()
+
+// auditResponseWriter intercepts WriteHeader, sets it in the event. If the sink is set, it will
+// create immediately an event (for long running requests).
+type auditResponseWriter struct {
+	http.ResponseWriter
+	event *auditinternal.Event
+	once  sync.Once
+	sink  chan<- *auditinternal.Event
+}
+
+func (a *auditResponseWriter) processCode(code int) {
+	a.once.Do(func() {
```

Comment:
huh.  Ok, so we're only writing *after* we complete processing?  That surprises me since irreversible action can be taken before we're called.  Why aren't we tracking beginning and end separately.  If a particular sink wants lower reliability they could choose to implement this logic themselves, right?